Code/Mantid/Framework/Algorithms/src/ExtractSpectra.cpp

```diff
@@ -0,0 +1,494 @@
+/** Executes the algorithm
+ *  @throw std::out_of_range If a property is set to an invalid value for the
+ * input workspace
+ */
+void ExtractSpectra::exec() {
+  // Get the input workspace
+  m_inputWorkspace = getProperty("InputWorkspace");
+
+  eventW = boost::dynamic_pointer_cast<EventWorkspace>(m_inputWorkspace);
+  if (eventW != NULL) {
+    // Input workspace is an event workspace. Use the other exec method
+    this->execEvent();
+  } else {
+    // Otherwise it's a Workspace2D
+    this->execHistogram();
+  }
+}
+
+/// Execute the algorithm in case of a histogrammed data.
+void ExtractSpectra::execHistogram() {
+  m_histogram = m_inputWorkspace->isHistogramData();
+  // Check for common boundaries in input workspace
+  m_commonBoundaries = WorkspaceHelpers::commonBoundaries(m_inputWorkspace);
+
+  // Retrieve and validate the input properties
+  this->checkProperties();
+
+  // Create the output workspace
+  MatrixWorkspace_sptr outputWorkspace = WorkspaceFactory::Instance().create(
+    m_inputWorkspace, m_spectrumList.size(), m_maxX - m_minX,
+      m_maxX - m_minX - m_histogram);
+
+  // If this is a Workspace2D, get the spectra axes for copying in the spectraNo
+  // later
+  Axis *inAxis1(NULL), *outAxis1(NULL);
+  TextAxis *outTxtAxis(NULL);
+  if (m_inputWorkspace->axes() > 1) {
+    inAxis1 = m_inputWorkspace->getAxis(1);
+    outAxis1 = outputWorkspace->getAxis(1);
+    outTxtAxis = dynamic_cast<TextAxis *>(outAxis1);
+  }
+
+  cow_ptr<MantidVec> newX;
+  if (m_commonBoundaries) {
+    const MantidVec &oldX = m_inputWorkspace->readX(m_spectrumList.front());
+    newX.access().assign(oldX.begin() + m_minX, oldX.begin() + m_maxX);
+  }
+  Progress prog(this, 0.0, 1.0, (m_spectrumList.size()));
+  // Loop over the required spectra, copying in the desired bins
+  for (int j = 0; j < static_cast<int>(m_spectrumList.size()); ++j) {
+    auto i = m_spectrumList[j];
+    // Preserve/restore sharing if X vectors are the same
+    if (m_commonBoundaries) {
+      outputWorkspace->setX(j, newX);
+    } else {
+      // Safe to just copy whole vector 'cos can't be cropping in X if not
+      // common
+      outputWorkspace->setX(j, m_inputWorkspace->refX(i));
+    }
+
+    const MantidVec &oldY = m_inputWorkspace->readY(i);
+    outputWorkspace->dataY(j)
+        .assign(oldY.begin() + m_minX, oldY.begin() + (m_maxX - m_histogram));
+    const MantidVec &oldE = m_inputWorkspace->readE(i);
+    outputWorkspace->dataE(j)
+        .assign(oldE.begin() + m_minX, oldE.begin() + (m_maxX - m_histogram));
+
+    // copy over the axis entry for each spectrum, regardless of the type of
+    // axes present
+    if (inAxis1) {
+      if (outAxis1->isText()) {
+        outTxtAxis->setLabel(j, inAxis1->label(i));
+      } else if (!outAxis1->isSpectra()) // handled by copyInfoFrom line
+      {
+        dynamic_cast<NumericAxis *>(outAxis1)
+            ->setValue(j, inAxis1->operator()(i));
```

Comment:
You are right, coverity should complain here but not at line 135 (I think). Unless it's so clever that it can figure it all out.